polkadot/xcm/xcm-builder/src/routing.rs

```diff
@@ -139,3 +139,69 @@ impl EnsureDelivery for Tuple {
 		(None, None)
 	}
 }
+
+/// A wrapper router that attempts to *encode* and *decode* passed XCM `message` to ensure that the
+/// receiving side will be able to decode, at least with the same XCM version.
+///
+/// This is designed to be at the top-level of any routers which do the real delivery. While other
+/// routers can manipulate the `message`, we cannot access the final XCM due to the generic
+/// `Inner::Ticket`. Therefore, this router aims to validate at least the passed `message`.
+///
+/// NOTE: For use in mock runtimes which don't have the DMP/UMP/HRMP XCM validations.
+pub struct EnsureDecodableXcm<Inner>(sp_std::marker::PhantomData<Inner>);
+impl<Inner: SendXcm> SendXcm for EnsureDecodableXcm<Inner> {
+	type Ticket = Inner::Ticket;
+
+	fn validate(
+		destination: &mut Option<Location>,
+		message: &mut Option<Xcm<()>>,
+	) -> SendResult<Self::Ticket> {
+		if let Some(msg) = message {
+			if let Some(e) = Self::ensure_encode_decode(msg) {
+				return Err(e)
+			}
+		}
+		Inner::validate(destination, message)
+	}
+
+	fn deliver(ticket: Self::Ticket) -> Result<XcmHash, SendError> {
+		Inner::deliver(ticket)
+	}
+}
+impl<Inner> EnsureDecodableXcm<Inner> {
+	fn ensure_encode_decode(xcm: &Xcm<()>) -> Option<SendError> {
+		use parity_scale_codec::Decode;
+		let encoded = xcm.encode();
+		match Xcm::<()>::decode(&mut &encoded[..]) {
+			Ok(decoded_xcm) => match decoded_xcm.eq(xcm) {
```

Comment:
It needs to check the nesting limit for encoded XCM.

Reply:
fixed by reused with `validate_xcm_nesting`, but moved to the separate PR: https://github.com/paritytech/polkadot-sdk/pull/4236